### ISSUE TITLE
Non-deterministic test results with nodejs and builtin providers

### ISSUE THREAD

**Issue by tsanders-rh** [konveyor/analyzer-lsp#1045]
# Non-Deterministic Test Results in Konveyor Analyzer

## Summary

Running the same test file multiple times produces different pass/fail results, indicating race conditions or non-deterministic behavior in the analyzer providers.

## Environment

- **Platform**: macOS (Darwin 25.1.0)
- **Kantra**: Latest version from ~/Workspace/kantra/bin/kantra
- **Test File**: `patternfly-v5-to-patternfly-v6-component-props.test.yaml`
- **Rule File**: `patternfly-v5-to-patternfly-v6-component-props.yaml` (41 rules)
- **Providers Used**: `nodejs` and `builtin`

## Reproduction Steps

1. Create a test with 41 rules using both `nodejs.referenced` and `builtin.filecontent` conditions
2. Run the same test 3 times consecutively without changing any files:
   ```bash
   cd tests
   kantra test patternfly-v5-to-patternfly-v6-component-props.test.yaml
   kantra test patternfly-v5-to-patternfly-v6-component-props.test.yaml
   kantra test patternfly-v5-to-patternfly-v6-component-props.test.yaml
   ```

## Expected Behavior

The same test file should produce identical results on every run when the source code and rules haven't changed.

## Actual Behavior

Test results vary randomly across runs:

```
Run 1: 28/41 passing (68.29%)
Run 2: 27/41 passing (65.85%)
Run 3: 26/41 passing (63.41%)
```

Different individual rules pass/fail on each run. For example:
- Rule `00050` passes in run 1, fails in run 2, passes in run 3
- Rule `00140` fails in run 1, passes in run 2, fails in run 3

## Example Rule Pattern

Most failing rules use this pattern structure:

```yaml
when:
  and:
    - nodejs.referenced:
        pattern: ComponentName
    - builtin.filecontent:
        pattern: <ComponentName[^>]*\bpropName\b
        filePattern: \.(j|t)sx?$
```

## Test Data Example

```tsx
import { Page, Masthead } from '@patternfly/react-core';

// Rule patternfly-v5-to-patternfly-v6-component-props-00050
<Page header={<Masthead />} />
```

Rule pattern: `<Page[^>]*\bheader\b`
- Sometimes matches → test passes
- Sometimes doesn't match → test fails
- Code never changes between runs

## Impact

- Cannot reliably validate rule correctness
- Test suites are unreliable for CI/CD
- Makes it difficult to debug actual rule issues vs. provider issues
- Blocks submission of rules because tests may pass locally but fail in CI

## Hypothesis

Possible causes:
1. **Race conditions**: Concurrent provider execution may have timing-dependent behavior
2. **Caching issues**: Provider cache may not invalidate correctly between runs
3. **File system race**: Reading files while provider is analyzing
4. **Provider initialization**: nodejs and builtin providers may not synchronize properly

## Related Issues

- This is separate from #1043 (multiline grep issue)
- This affects patterns that DO work with grep when they trigger

## Additional Context

The issue appears when:
- Multiple rules test the same file
- Rules use `and` conditions combining `nodejs.referenced` + `builtin.filecontent`
- Test file has 40+ rules (may be related to concurrency/load)

Simpler test files with 2-5 rules appear to have consistent results.

## Workaround

None known. Running tests multiple times and accepting "best result" is not viable for production use.

**Comment by tsanders-rh** [comment]
## Root Cause Analysis

I've investigated this issue and identified **3 critical race conditions** causing the non-deterministic test behavior with `nodejs.referenced` + `builtin.filecontent` AND conditions.

### 1. 🔴 CRITICAL: Location Cache Defer Bug

**File:** `provider/internal/builtin/service_client.go:424-437`

**The Problem:**

The defer statement writes to cache **before** the value is computed:

```go
defer func() {
    b.cacheMutex.Lock()
    b.locationCache[cacheKey] = location.StartPosition.Line
    b.cacheMutex.Unlock()
}()

location.StartPosition.Line = -1  // Line 430
lineNumber, err := provider.MultilineGrep(ctx, len(lines), path, pattern)
// ... MultilineGrep computation happens here ...
location.StartPosition.Line = float64(lineNumber)  // Line 435
```

**Impact:**
- Defer captures the cache write immediately at function entry
- Can execute before `lineNumber` is computed
- Concurrent rules reading the same file get `-1` or intermediate values
- Directly causes wrong line numbers and missed matches in `builtin.filecontent` patterns

**Evidence:** This affects XML/JSON location lookups, which explains why rules with `builtin.filecontent` patterns randomly fail.

---

### 2. 🟡 HIGH: File Walking Cache Not Thread-Safe

**File:** `provider/lib.go:293-348`

**The Problem:**

```go
func newCachedWalkDir() cachedWalkDir {
    cache := make(map[string]struct {
        files []string
        err   error
    })
    return func(basePath string, ...) {
        val, ok := cache[basePath]  // ⚠️ RACE: concurrent read
        if !ok {
            // ...
            cache[basePath] = struct {...}{...}  // ⚠️ RACE: concurrent write
        }
        return val.files, val.err
    }
}
```

**Impact:**
- No mutex protection on map access
- Classic Go map race condition
- Can panic or return incorrect file lists with 10 concurrent workers

---

### 3. 🟠 MEDIUM: Concurrent Rule Processing Amplifies Issues

**File:** `engine/engine.go:135-158`

**Context:**
- Engine creates 10 worker goroutines that process rules concurrently
- With 40+ rules, creates high cache contention
- All race conditions manifest simultaneously

---

## Why This Manifests with AND Conditions

When rules use:
```yaml
when:
  and:
    - nodejs.referenced: ...
    - builtin.filecontent: ...
```

The sequence creates a race window:
1. `nodejs` provider evaluates (may trigger file processing)
2. `builtin` provider evaluates (reads from potentially corrupted cache)
3. Cache state depends on timing of 10 concurrent workers
4. Results vary based on which goroutine wins the race

---

## Reproduction Pattern

✅ **Triggers the bug:**
- 40+ rules (high concurrency)
- Same files analyzed by multiple rules
- `and` conditions crossing provider boundaries
- `builtin.filecontent` patterns (uses location cache)

❌ **Doesn't trigger:**
- 2-5 rules (low concurrency, less cache contention)
- Single provider rules (no cross-provider timing issues)
- Simple patterns without location lookup

This matches the observed behavior: 68.29%, 65.85%, 63.41% pass rates across identical runs.

---

## Recommended Fixes (Priority Order)

### Priority 1: Fix Location Cache Defer Bug

**Current (BROKEN):**
```go
defer func() {
    b.cacheMutex.Lock()
    b.locationCache[cacheKey] = location.StartPosition.Line
    b.cacheMutex.Unlock()
}()
location.StartPosition.Line = -1
// ... compute lineNumber ...
location.StartPosition.Line = float64(lineNumber)
```

**Fixed:**
```go
location.StartPosition.Line = -1
lineNumber, err := provider.MultilineGrep(ctx, len(lines), path, pattern)
// ... handle error ...
location.StartPosition.Line = float64(lineNumber)

// Write to cache AFTER computation
b.cacheMutex.Lock()
b.locationCache[cacheKey] = location.StartPosition.Line
b.cacheMutex.Unlock()
```

### Priority 2: Make File Walking Cache Thread-Safe

Add mutex protection:
```go
func newCachedWalkDir() cachedWalkDir {
    cache := make(map[string]struct {
        files []string
        err   error
    })
    var mu sync.RWMutex  // ADD THIS

    return func(basePath string, ...) {
        mu.RLock()
        val, ok := cache[basePath]
        mu.RUnlock()

        if !ok {
            // ... walk files ...
            mu.Lock()
            cache[basePath] = struct {...}{files: files, err: err}
            mu.Unlock()
        }
        return val.files, val.err
    }
}
```

### Priority 3: Share Location Cache Across Service Clients

Currently each `builtinServiceClient` has its own `locationCache`. Consider:
- **Option A:** Use a single shared cache with proper locking
- **Option B:** Move cache to provider level instead of client level
- **Option C:** Use `sync.Map` for lock-free concurrent access

---

## Testing Notes

**Note:** I attempted to reproduce this with unit tests (`Test_ConcurrentLocationCacheDeferRace` with 100 concurrent goroutines), but the race didn't manifest because:

1. Simple unit tests have high cache hit rates (first goroutine caches, rest hit cache)
2. Small test files complete MultilineGrep in microseconds
3. This is a **logic race** (wrong values cached) not a **data race** (concurrent map access)

The Go race detector doesn't catch "wrong value cached at wrong time" logic errors - it only catches concurrent memory access violations. This is why the issue manifests as non-deterministic test results rather than race detector warnings.

**To verify fixes:** Run the actual failing test case (patternfly-v5-to-patternfly-v6-component-props.test.yaml) 10 times and confirm consistent results.

---

## Files to Examine

1. `provider/internal/builtin/service_client.go:424-437` (location cache bug)
2. `provider/lib.go:293-348` (file walking cache)
3. `engine/engine.go:135-158` (concurrent rule processing)
4. `provider/internal/builtin/provider.go:122-143` (client initialization)

Happy to provide more details or help with testing the fixes!

**Comment by shawn-hurley** [comment]
nodejs. referenced and built-in. filecontent or other conditions that in the same rule will run sequentially and not asynchronously, so I am having a hard time seeing this being a race condition. 

I have a feeling this was fixed by the preparation work @pranavgaikwad put in for the NodeJS provider, and since @mguetta1 tested it and couldn't reproduce, I think we can close. 

If you are still seeing this issue, please reopen.
